File: jdroid-android-sample/src/main/java/com/jdroid/android/sample/androidx/SampleViewModel.java
package com.jdroid.android.sample.androidx;

import com.jdroid.android.androidx.lifecycle.Resource;
import com.jdroid.android.androidx.viewmodel.AbstractViewModel;
import com.jdroid.android.sample.database.room.SampleEntity;

import androidx.lifecycle.LiveData;

public class SampleViewModel extends AbstractViewModel {

	private LiveData<Resource<SampleEntity>> sampleEntity;

	private Boolean failLoadFromNetwork = false;
	private Boolean failLoadFromDb = false;
	private Boolean failSaveToDb = false;
	private Integer loadFromNetworkDelaySeconds = 0;
	private Integer loadFromDbDelaySeconds = 0;
	private Integer saveToDbDelaySeconds = 0;

	public LiveData<Resource<SampleEntity>> load(String id, Boolean forceRefresh) {
		if (sampleEntity == null || forceRefresh) {
			sampleEntity = SampleRepository.get(id, forceRefresh, failLoadFromNetwork, failLoadFromDb, failSaveToDb, loadFromNetworkDelaySeconds, loadFromDbDelaySeconds, saveToDbDelaySeconds);
		}
		return sampleEntity;
	}

	public LiveData<Resource<SampleEntity>> getSampleEntity() {
		return sampleEntity;
	}

	public void setFailLoadFromNetwork(Boolean failLoadFromNetwork) {
		this.failLoadFromNetwork = failLoadFromNetwork;
	}

	public void setFailLoadFromDb(Boolean failLoadFromDb) {
		this.failLoadFromDb = failLoadFromDb;
	}

	public void setFailSaveToDb(Boolean failSaveToDb) {
		this.failSaveToDb = failSaveToDb;
	}

	public void setLoadFromNetworkDelaySeconds(Integer loadFromNetworkDelaySeconds) {
		this.loadFromNetworkDelaySeconds = loadFromNetworkDelaySeconds;
	}

	public void setLoadFromDbDelaySeconds(Integer loadFromDbDelaySeconds) {
		this.loadFromDbDelaySeconds = loadFromDbDelaySeconds;
	}

	public void setSaveToDbDelaySeconds(Integer saveToDbDelaySeconds) {
		this.saveToDbDelaySeconds = saveToDbDelaySeconds;
	}
}

File: jdroid-android-sample/src/main/java/com/jdroid/android/sample/androidx/SampleViewModel.kt
package com.jdroid.android.sample.androidx

import androidx.lifecycle.LiveData
import com.jdroid.android.androidx.lifecycle.Resource
import com.jdroid.android.androidx.viewmodel.AbstractViewModel
import com.jdroid.android.sample.database.room.SampleEntity

class SampleViewModel : AbstractViewModel() {

    var sampleEntity: LiveData<Resource<SampleEntity>>? = null
        private set

    private var failLoadFromNetwork: Boolean = false
    private var failLoadFromDb: Boolean = false
    private var failSaveToDb: Boolean = false
    private var loadFromNetworkDelaySeconds: Int = 0
    private var loadFromDbDelaySeconds: Int = 0
    private var saveToDbDelaySeconds: Int = 0

    fun load(id: String, forceRefresh: Boolean): LiveData<Resource<SampleEntity>> {
        if (sampleEntity == null || forceRefresh) {
            sampleEntity = SampleRepository.getItem(
                id,
                forceRefresh,
                failLoadFromNetwork,
                failLoadFromDb,
                failSaveToDb,
                loadFromNetworkDelaySeconds,
                loadFromDbDelaySeconds,
                saveToDbDelaySeconds
            )
        }
        return sampleEntity!!
    }

    fun setFailLoadFromNetwork(failLoadFromNetwork: Boolean) {
        this.failLoadFromNetwork = failLoadFromNetwork
    }

    fun setFailLoadFromDb(failLoadFromDb: Boolean) {
        this.failLoadFromDb = failLoadFromDb
    }

    fun setFailSaveToDb(failSaveToDb: Boolean) {
        this.failSaveToDb = failSaveToDb
    }

    fun setLoadFromNetworkDelaySeconds(loadFromNetworkDelaySeconds: Int) {
        this.loadFromNetworkDelaySeconds = loadFromNetworkDelaySeconds
    }

    fun setLoadFromDbDelaySeconds(loadFromDbDelaySeconds: Int) {
        this.loadFromDbDelaySeconds = loadFromDbDelaySeconds
    }

    fun setSaveToDbDelaySeconds(saveToDbDelaySeconds: Int) {
        this.saveToDbDelaySeconds = saveToDbDelaySeconds
    }
}
